### ISSUE TITLE
Large generated parser bumps against compiler size limit

### ISSUE THREAD

**Issue by arnt** [phax/ParserGeneratorCC#28]
com.helger.css.parser.ParserCSS30TokenManager is problematic on Android, because the CSS grammar is too large for the JIT compiler. I see other people have run similar problems, e.g. [this htmlunit issue](https://github.com/HtmlUnit/htmlunit-cssparser/issues/4).

Having cast the merest glance at the source, I don't see a way to do it, but it would be lovely if ParserGeneratorCC were to produce many smallish methods, so that no one method would scale in size with the grammar.

**Comment by phax** [comment]
@arnt thx for the issue.
The issue you are mentioning (HtmlUnit/htmlunit-cssparser#4) refers to JaCoCo in CodeCoverage analysis. You are mentioning Android. Do you mean Code Coverage on Android? Can you paste an error message here?

**Comment by arnt** [comment]
This is with ph-css 6.2.3: `Method exceeds compiler instruction limit: 40516 in int com.helger.css.parser.ParserCSS30TokenManager.jjMoveNfa_0(int, int)` and it seems to happen whenever I call either `CSSReaderDeclarationList.readFromString()` or `CSSReader.readFromString()`.

The issue affects android worse than real-java, because android has a JIT that gives up and complains if a single method contains more than 16383 JVM opcodes, while real-java accepts 65535 opcodes in a single method. But I think it's a general issue, not android-specific. 40516 is too close to 65535 for comfort.

**Comment by phax** [comment]
Okay, as Android is not my strong side, it seems to be comaprable to this:
https://stackoverflow.com/questions/54960389/method-exceeds-compiler-instruction-limit-message-on-well-under-64k-method
If the limit is "16383.75" it will be tough as 40516 is way above it ;-)

For what Android SDK version are you programming?

**Comment by arnt** [comment]
Yes, that's someone else with a big method.

I think the core problem is that the jjMoveNfs_0 method is 5689 lines long in the case of CSS30. You could say that the problem is that Android's size limit is unreasonably small, but generating 5689 lines in one method for such a small grammar is sailing close to the wind. Tools that process .class files *should* accept up to 65535 opcodes. That is often around 8-10k source lines, 5689 is close to that.

I chose to report this as a PGCC issue because 5689 lines for a small grammar like that of CSS implies that other grammars could easily pass 8-10k.

The Android limit applies to every version since.... 5.0, I think. Maybe 5.1.

(If anyone with Android problems finds this issue: I was able to minimise the problem using `Map<String,CascadingStyleSheet>` and `Map<String,CSSDeclarationList>`. This works quite well in my case, because I see many style=foo attributes but only a few different values of foo, so the JIT compiler only burns CPU/battery a few times instead of many.)

**Comment by phax** [comment]
Yes, I got that. Unfortunately that is really the heart of the JavaCC code generation and the code quality is *very* poor and the complexity is high because of all the static variables that are used.
Any suggestions how to simplify this would be welcome....

**Comment by arnt** [comment]
Looking again, it seems to me that moving some of the switch() cases into subroutines might help. Perhaps move a case into a function of its own if it would add more than two lines to jjMoveNfa_0, or move all of them if there are more than 200 cases in the switch. Something like that.

It's not a simplification though. I think if I were you I'd close this issue as hopeless.

**Comment by revusky** [comment]
This whole problem was solved pretty fundamentally in JavaCC 21 almost two years ago. See: https://javacc.com/2021/03/30/code-too-large-problem-fixed/
